### ISSUE TITLE
Perhaps we should change the back-buffer size now?

### ISSUE THREAD

**Issue by MatthewKush** [xrOxygen/xray-oxygen#190]
Hi, so I am writing this here because I want to know how you guys feel about this.

Basically, my thinking is that I would assume that almost nobody has a 4:3 monitor at this point, and if they still do my guess is they can achieve better than 1024x768 resolution at this point (however I'm a bit naive about how things are in other parts of the world, I need input).

Basically, in xrEngine/defines.cpp is where the initial backbuffer size is defined as being 1024x768. Why not change this to 1280x720 by default, then add a check to see if the user has a 1080p monitor: if so, the backbuffer size is set to 1920x1080 by default. And of course for "Native 4k" the backbuffer could be set to that.

To get to the point: this has an actual impact on how the game image looks believe it or not (if I am wrong about this please correct me, but I can definitely tell the difference). The higher the resolution of the initial backbuffer, the better the final image quality.

Obviously though, you would not want to set the backbuffer size to something that is not supported by the end user's monitor.

Any ideas, is my thinking wrong, etc? I'd especially like to hear from @RainbowZerg as he seems to have a better understanding of rendering than I do.

P.S. I just want to state that I agree to adhere to the Code of Conduct.

**Comment by RainbowZerg** [comment]
Hi, Matthew. I think that it would be nice to remove these fixed values from xrEngine/defines.cpp and select initial resolution based on the user's desktop resolution like it's done in most of modern games.
I'm not sure that i correctly understood your words about "The higher the resolution of the initial backbuffer, the better the final image quality" so the following might be obvious for you. Back buffers are recreated after each d3d device reset and their resolution is the same as selected by the user in options, so initial resolution will be used only in that case when game was started without user.ltx or there is not "vid_mode" line in the file.

**Comment by MatthewKush** [comment]
@RainbowZerg Oh ok. I guess I did not understand that correctly then. I'm just curious then: these games that are claiming to be "native 4k", what does that even mean then?

P.S. I love the work you're doing. I'm learning more about rendering from your work.

**Comment by RainbowZerg** [comment]
@MatthewKush I don't know what "native 4k" really means, perhaps it is about consoles, not all of which supports this resolution (some of them can only render at 1080p and then stretching image to fit 4k screen), and/or about game's textures size.
On a PC, if the game correctly enumerates supported display modes (what does not always happen to old games which even haven't a widescreen support), all you have to play in 4k resolution is a video card that supports such back buffer size and a suitable monitor, although there can be some glitches like a stretched UI if developers originally did not plan to support such resolution.

P.S. Thanks. I also learned some new about rendering from your commits.

**Comment by Xottab-DUTY** [comment]
Actually, a better name for `psCurrentVidMode` is `wantedResolution`. That is how this variable is used in the engine and it's default value (1024,768) should have no impact on final image quality. Otherwise this should be fixed. Probable cause of this is that it may be used before the actual resolution selection.

**Comment by MatthewKush** [comment]
@Xottab-DUTY Yep, exactly, they set it as 1024 and 768 as the initial size as that was a common resolution at the time and it was a good initial value for the backbuffer. I guess what I misunderstood is I guess I thought that somehow by using a larger initial image it somehow improved image quality, sort of like super-sampling, but I have obviously misunderstood that.

Speaking of super-sampling, might I propose just that as an option? I've never been a fan of MSAA, especially since Stalker uses a Deffered renderer so MSAA runs slow. FXAA and SMAA are great, but I know that personally, even when I'm using whatever form of AA, if a game has "resolution scaling" as an option I always use it and usually turn it on to 1.25 - 1.5. 

I think it works the same way as super-sampling, but maybe there's a reason it's being given a different name, like maybe before people only used 2x or 4x super-sampling and now simply using stuff like 1.15-1.5 is becoming common place. I'm not 100% on this though.

**Comment by Xottab-DUTY** [comment]
@MatthewKush, hi Matty! What do you think about changing the backbuffer format from 32-bit R8G8B8A8 to 64-bit R16G16B16A16? I tried to do this and got gamma overexposure. The image on the screen was too white.

**Comment by MatthewKush** [comment]
@MatthewKush Hi. Yeah you can do that, but then you'll need to play with the tonemapper values and sun_lumscale values to get things back into the appropriate range again. It could be a lot of work that way.
Instead I recommend using the current industry-standard: R10G10B10A2. With this format you'll also have HDR-10 support, though obviously common-usage of HDR-10 is still a long way off (the monitors that have it are absurdly expensive, I don't recommend anybody buy them for a long, long time).

However, another benefit would more precise colors.

I've tried changing the back buffer to this format multiple times on this repository, but whenever I do at the main menu my mouse cursor is invisible. I seem to remember it working on your open-xray repo though.

BTW, try out this new tonemapper (Insomniac Game's Filmic tonemapper):

float	ToneMappingFilmic_Insomniac( float x ) {
	float	w = 1.7;//0.665;	// White point
	float	b = 0.09;				// Black point
	float	c = 0.004;				// Junction point
	float	t = 0.445;				// Toe strength
	float	s = 0.32;				// Shoulder strength
	float	k = 0.02;				// Junction factor
	#if 0
		// Simple version
		return	x < c
			? k*(1-t)*(x-b) / ((c-b) - t*(x-b))
			: k+(1-k)*(x-c) / ((1-s)*(w-c) + s*(x-c));
	#else
		// "Optimized" version where coeffs for Toe and Shoulder could be passed as float4 to the CB
		float4	Coeffs_Toe = float4( k-k*t, -t, -k*b + k*b*t, c-b+t*b );
		float4	Coeffs_Shoulder = float4( k*(s-1)+1, s, k*(1-s)*(w-c) - k*s*c - c*(1-k), (1-s)*(w-c) - s*c );

		float4	Coeffs = x < c ? Coeffs_Toe : Coeffs_Shoulder;
		float2	Fraction = Coeffs.xy * x + Coeffs.zw;
		return Fraction.x / Fraction.y;
	#endif
}
float3	ToneMappingFilmic_Insomniac( float3 x ) {
	return float3( ToneMappingFilmic_Insomniac( x.x ), ToneMappingFilmic_Insomniac( x.y ), ToneMappingFilmic_Insomniac( x.z ) );
}
void tonemap( out float4 low, out float4 high, float3 rgb, float scale)
{
	rgb		=	rgb*scale;

	low		=	ToneMappingFilmic_Insomniac(rgb).xyzz;

	high	=	rgb.xyzz/def_hdr;	// 8x dynamic range
}

You'll want to play around with the different values to find what you like. I used the white point from the original tonemapper Stalker uses (which is a modified Reinhard).

**Comment by Xottab-DUTY** [comment]
@MatthewKush, yes `R10G10B10A2` is working fine in OpenXRay, except one thing – when you're trying to make a screenshot, it fails. The same appears to be in the Oxygen. Switch to DirectTex from D3DX should fix this probably.
